scripts/bigquery.py

```diff
@@ -0,0 +1,392 @@
+
+
+def run(command: Union[str, List[str]], **kwargs) -> str:
+    """Simple wrapper around subprocess.run that returns stdout and raises exceptions on errors."""
+    if isinstance(command, list):
+        args = command
+    elif isinstance(command, str):
+        args = command.split()
+    else:
+        raise RuntimeError(f"run command is invalid: {command}")
+
+    # TODO: log the output
+    return (
+        subprocess.run(args, stdout=subprocess.PIPE, **{**dict(check=True), **kwargs})
+        .stdout.decode()
+        .strip()
+    )
+
+
+def transpile(schema_path: Path) -> dict:
+    """Transpile a JSON Schema into a BigQuery schema."""
+    res = run(
+        [
+            "jsonschema-transpiler",
+            str(schema_path),
+            "--normalize-case",
+            "--resolve",
+            "cast",
+            "--type",
+            "bigquery",
+        ]
+    )
+    schema = json.loads(res)
+    return schema
+
+
+def transform(document: dict) -> dict:
+    """Transform a document for loading into BigQuery."""
+    # TODO: normalize field names using snake casing
+    # TODO: additional properties
+    # TODO: pseudo maps
+    # TODO: anonymous structs
+    raise NotImplementedError()
+
+
+def transpile_schemas(output_path: Path, schema_paths: List[Path]):
+    """Write schemas to directory."""
+    assert output_path.is_dir()
+    for path in schema_paths:
+        namespace, doctype, filename = path.parts[-3:]
+        version = int(filename.split(".")[-3])
+        # pioneer-study schemas were done incorrectly and are ignored here
+        if namespace == "schemas":
+            print(f"skipping {path} due to wrong directory level")
+            continue
+        out = output_path / f"{namespace}.{doctype}.{version}.bq"
+        with out.open("w") as fp:
+            print(f"writing {out}")
+            json.dump(transpile(path), fp, indent=2)
+            fp.write("\n")
+
+
+def load_schemas(input_path: Path):
+    """Load schemas into memory for use in google-cloud-bigquery."""
+    paths = list(input_path.glob("*.bq"))
+    assert len(paths) > 0
+    schemas = {}
+    for path in paths:
+        qualified_name = path.parts[-1][:-3]
+        with path.open("r") as fp:
+            schemas[qualified_name] = json.load(fp)
+    print(f"loaded {len(schemas.keys())} schemas")
+    return schemas
+
+
+def git_stash_size():
+    """Find the size of the git stash."""
+    return len([item for item in run("git stash list").split("\n") if item])
+
+
+def resolve_ref(ref: str) -> str:
+    """Return a resolved reference or the short revision if empty."""
+    resolved = run(f"git rev-parse --abbrev-ref {ref}") or run(
+        f"git rev-parse --short {ref}"
+    )
+    if resolved != ref:
+        print(f"resolved {ref} to {resolved}")
+    return resolved
+
+
+@contextmanager
+def managed_git_state():
+    """Save the current git state.
+
+    Stash any changes so we can reference by real changes in the tree. If the
```

Comment:
This is true, I think it's probably best resolved by failing if the branch is unclean in the templates/schemas directory. I'd prefer not to commit changes for the user to avoid any user input during the script.

Reply:
Got it, makes sense. Most people will wait for CI to verify these changes anyways.

Reply:
I've added an assertion on `git_untracked_files` to prevent this state from occuring.